x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IndexLifecycleRunner.java

```diff
@@ -504,4 +507,38 @@ void registerFailedOperation(IndexMetadata indexMetadata, Exception failure) {
                 LifecycleExecutionState.fromIndexMetadata(indexMetadata),
                 failure));
     }
+
+    private final Set<IndexLifecycleClusterStateUpdateTask> executingTasks = new HashSet<>();
+
+    /**
+     * Tracks already executing {@link IndexLifecycleClusterStateUpdateTask} tasks in {@link #executingTasks} to prevent queueing up
+     * duplicate cluster state updates.
+     * TODO: refactor ILM logic so that this is not required any longer. It is unreasonably expensive to only filter out duplicate tasks at
+     *       this point given how these tasks are mostly set up on the cluster state applier thread.
+     *
+     * @param source source string as used in {@link ClusterService#submitStateUpdateTask(String, ClusterStateTaskConfig)}
+     * @param task   task to submit unless already tracked in {@link #executingTasks}.
+     */
+    private void submitUnlessAlreadyQueued(String source, IndexLifecycleClusterStateUpdateTask task) {
+        if (registerTask(task)) {
+            task.addListener(ActionListener.wrap(() -> unregisterTask(task)));
+            clusterService.submitStateUpdateTask(source, task);
+        } else {
+            logger.trace("skipped redundant execution of [{}]", source);
+        }
+    }
+
+    private boolean registerTask(IndexLifecycleClusterStateUpdateTask task) {
+        synchronized (executingTasks) {
```

Comment:
any reason for explicit `synchronized` block and not using `Collections.synchronizedSet(new HashSet<>())` or `ConcurrentHashMap`?

Reply:
Right ... not sure why I did it this  used the sync set now and just inlined everything since it's all one liners now :)

Thanks!